lib/semantics/semantics.cc

```diff
@@ -208,6 +209,70 @@ void SemanticsContext::PopConstruct() {
   constructStack_.pop_back();
 }
 
+void SemanticsContext::SayDoVarRedefine(
+    const Symbol &variable, const parser::CharBlock &location) {
+  const parser::CharBlock doLoc{GetDoVariableLocation(variable)};
+  Say(location, "Cannot redefine DO variable %s"_err_en_US, variable.name())
+      .Attach(doLoc, "Enclosing DO construct"_en_US);
+}
+
+void SemanticsContext::CheckDoVarRedefine(
+    const Symbol &variable, const parser::CharBlock &location) {
+  if (const Symbol * root{GetAssociationRoot(variable)}) {
+    if (IsActiveDoVariable(*root)) {
+      SayDoVarRedefine(*root, location);
+    }
+  }
+}
+
+void SemanticsContext::CheckDoVarRedefine(const parser::Variable &variable) {
+  if (const Symbol * entity{GetLastName(variable).symbol}) {
+    const parser::CharBlock &sourceLocation{variable.GetSource()};
+    CheckDoVarRedefine(*entity, sourceLocation);
+  }
+}
+
+void SemanticsContext::CheckDoVarRedefine(const parser::Name &name) {
+  const parser::CharBlock &sourceLocation{name.source};
+  if (const Symbol * entity{name.symbol}) {
+    CheckDoVarRedefine(*entity, sourceLocation);
+  }
+}
+
+void SemanticsContext::ActivateDoVariable(const parser::Name &name) {
+  CheckDoVarRedefine(name);
+  if (const Symbol * doVariable{name.symbol}) {
+    if (const Symbol * root{GetAssociationRoot(*doVariable)}) {
+      if (!IsActiveDoVariable(*root)) {
+        activeDoVariables_.emplace(*root, name.source);
+      }
+    }
+  }
+}
+
+void SemanticsContext::DeactivateDoVariable(const parser::Name &name) {
+  if (Symbol * doVariable{name.symbol}) {
+    if (const Symbol * root{GetAssociationRoot(*doVariable)}) {
+      if (name.source == GetDoVariableLocation(*root)) {
+        activeDoVariables_.erase(*root);
+      }
+    }
+  }
+}
+
+bool SemanticsContext::IsActiveDoVariable(const Symbol &variable) {
+  return activeDoVariables_.find(variable) != activeDoVariables_.end();
+}
+
+parser::CharBlock SemanticsContext::GetDoVariableLocation(
+    const Symbol &variable) {
+  if (IsActiveDoVariable(variable)) {
```

Comment:
When would this be called on something that isn't an active do variable? If this actually did return an empty CharBlock to `SayDoVarRedefine` we wouldn't get a good error message location.

Reply:
GetDoVariableLocation() gets called in two situations.  One is to get the location for error messages.  The other is to deactivate a DO variable at the end of a DO loop.  In the case of nested DO loops that use the same variable, we only want to deactivate the loop variable for the outer loop when we exit it, not when we exit the inner loop.  So DeactivateDoVariable() calls GetDoVariableLocation() to make sure that its deactivating the correct variable.

I don't think GetDoVariableLocation() will be called for an error message when there's not an empty location for the DO variable.  But I can certainly add a check that will verify that.